feat(kdropdownmenu): new component [khcp-690] (#724)
Kong/kongponents: docs/components/dropdown-menu.md

> # DropdownMenu

Reviewer: Page title fix
```suggestion
# Dropdown Menu
```

Author: Fixed

**KDropdownMenu** is a button (or any slotted content) that is clicked to trigger a menu popover beneath it.

<KDropdownMenu label="Documentation" :items="deepClone(defaultItemsUnselected)" />

## Props

### label

The label for the menu.

### items

> An array of items containing a `label` and other optional properties will render a menu of [`KDropdownItems`](#KDropdownItem) .

Reviewer: ```suggestion
An array of item objects containing a `label` property and other optional properties which will render a menu of [`KDropdownItems`](#KDropdownItem) .
```

```html
<KDropdownMenu
  label="Documentation"
  :items="[
    { label: 'Home', to: { path: '/' } },
    { label: 'Button docs', to: { path: '/components/button.html' } },
    { label: 'My docs', to: { path: '/components/dropdown-menu.html' } }
  ]"
/>
```

### appearance

Use this prop to specify the display style for the dropdown menu. Can be either `menu` (default) or `selectionMenu`.

> The `menu` style is the standard you have seen in the example above.

Reviewer: Can you mention here that it renders a KButton style, etc?

Author: Done

The `selectionMenu` style is good for a clearer representation of what is selected. `selected` state is handled automatically when

> clicking a `KDropdownItem` if used in conjunction with the `items` prop. You will need to manually control selectedness if you are using the `items` slot.

Reviewer: "selectedness" can we reword this? to something like this? (Unsure if this reads correctly 😬)
```suggestion
clicking a `KDropdownItem` if used in conjunction with the `items` prop. 

If you instead utilize the `items` slot, you will need to manually control the reactive state of the selected item.
```

Author: Fixed (I think yours reads better than mine, I like to make up words 😓)

  appearance="selectionMenu"
>
  <template #items>
    <KDropdownItem
      :selected="selectedItem === 'us'"
      @click="clickHandler('US selected', 'us', 'US (United States)')"
    >
      US (United States)
    </KDropdownItem>
    <KDropdownItem
      :selected="selectedItem === 'fr'"
      @click="clickHandler('France selected', 'fr', 'FR (France)')"
    >
      FR (France)
    </KDropdownItem>
  </template>
</KDropdownMenu>

<script>
export default {
  data() {
    return {
      selectedItem: '',
      selectedLabel: 'Select an item'
    }
  },
  methods: {
    clickHandler (msg, val, label) {
      if (val !== undefined) {
        this.selectedItem = val
      }

      if (label) {
        this.selectedLabel = label
      }

      this.$toaster.open(msg)
    }
  }
}
</script>
```

### showCaret

Use this prop if you would like the trigger button to display the caret.

<KDropdownMenu label="Documentation" :items="deepClone(defaultItemsUnselected)" show-caret />

```html
<KDropdownMenu
  label="Documentation"
  :items="items"
  show-caret
/>
```

### kpopAttributes

Use the `kpopAttributes` prop to configure the **KPop** [props](/components/popover.html) dropdown.

> 

Reviewer: I'm not sure the card example is ideal here since the Docs user isn't familiar with the "three dot" action menu UI and may not understand this example without clicking around or some other instruction?

Author: Fixed - added instruction

![image](https://user-images.githubusercontent.com/67973710/182684007-202c7c8b-c1f3-406c-9a68-0d25c4f510bb.png)

                color="var(--black-400)"
                size="16"
              />
            </template>
          </KButton>
        </template>
      </KDropdownMenu>
    </template>
  </KCard>
</div>

```html
<KCard
  title="Card Title"
  body="Body Content"
>
  <template #actions>
    <KDropdownMenu
      :kpop-attributes="{
        popoverClasses: 'mt-5',
        width: '180'
      }"
      :items="deepClone(defaultItemsUnselected)"
    >
      <template #default>
        <KButton
          appearance="secondary"
          size="small"
          class="non-visual-button"
        >
          <template #icon>
            <KIcon
              icon="more"
              color="var(--black-400)"
              size="16"
            />
          </template>
        </KButton>
      </template>
    </KDropdownMenu>
  </template>
</KCard>
```

### disabled

Use this prop to disable the dropdown, can be used in conjunction with `disabledTooltip` prop.

### disabledTooltip

Text to display on hover if dropdown is disabled.

<div>
  <KDropdownMenu
    label="Documentation"
    :disabled="true"
    disabled-tooltip="You can't click me"
    :items="deepClone(defaultItemsUnselected)"
  />

> </div>

Reviewer: issue: I can hover/click on the dropdown, the tooltip shows. If I then click on the page, the dropdown itself shows.

Reviewer: The width of the element itself is too large and allows you to click (even not on the button) 
![image](https://user-images.githubusercontent.com/2229946/182469745-a170ce3d-9a2f-4f6a-8d09-4188eafe9ea6.png)

Author: Fixed

```html
<KDropdownMenu
  label="Documentation"
  :disabled="true"
  disabled-tooltip="You can't click me"
  :items="deepClone(defaultItemsUnselected)"
/>
```

## Slots

There are 2 supported slots:

> - `default` - The trigger element for opening/closing the menu. Returns the `isOpen` state.

Reviewer: I'm not sure what this means? I think you're referring to the scoped slot data? 

> Returns the `isOpen` state

Author: Updated the text: "Slot provides `isOpen` - whether the menu is open or not."

- `items` - For an example of using the items slot see the [`KDropdownItem`](#KDropdownItem) section.

<div>

>   <KDropdownMenu :items="deepClone(defaultItemsUnselected)">

Reviewer: Similar to the changes we made on some other components, can we update the examples to not link away from the current page (eg linking to the button docs)? Maybe just show a tooltip or something instead?

Author: So, in order to have actionable items using the `items` prop, they have to be links. I compromised by only linking to anchor tags within the current page.

    <template #default="{ isOpen }">
        <KButton
          :is-open="isOpen"
          appearance="creation"
        >
          Menu
        </KButton>
      </template>
  </KDropdownMenu>
</div>

```html
<KDropdownMenu :items="items">
  <template #default="{ isOpen }">
      <KButton
        :is-open="isOpen"
        appearance="creation"
      >
        Menu
      </KButton>
    </template>
</KDropdownMenu>
```

## KDropdownItem

> **KDropdownMenu** generates a **KDropdownItem** for each item in the `items` property. At the most basic level, **KDropdownItem** is a wrapper around each item to display it correctly inside `KDropdownMenu`. You can use the `items` slot of the `KDropdownMenu` to manually create your own menu items.

Reviewer: Just a little rewording 
```suggestion
**KDropdownMenu** generates a **KDropdownItem** for each object in the `items` prop array. At the most basic level, **KDropdownItem** is a wrapper around each item to display it correctly inside `KDropdownMenu`. You can use the `items` slot of the `KDropdownMenu` to manually create your own menu items.
```

### Properties

> - `item` - the properties the router-link is built from, it expects a `label` and a `to`.

Reviewer: See my comment about router-links -- ideally I think we should require using the slot for router-links so that we don't have a Vue router peer dependency

> - `disabled` - a boolean (default to `false`), whether or not to disable the item.

Reviewer: ```suggestion
- `disabled` - a boolean (defaults to `false`), whether or not to disable the item.
```

> - `selected` - a boolean (default to `false`), whether or not the item is selected.

Reviewer: ```suggestion
- `selected` - a boolean (defaults to `false`), whether or not the item is selected.
```

> - `hasDivider` - a boolean (default to `false`), whether or not the item should have a divider bar displayed above it

Reviewer: ```suggestion
- `hasDivider` - a boolean (defaults to `false`), whether or not the item should have a divider bar displayed above it
```

> - `isDangerous` - a boolean (default to `false`), whether or not to apply danger styles (text color is red)

Reviewer: ```suggestion
- `isDangerous` - a boolean (defaults to `false`), whether or not to apply danger styles (text color is red)
```

Also see comment about this prop name

> - `selectionMenuChild` - a boolean (default to `false`), whether the parent is a `selectionMenu` or not (used for events)

Reviewer: ```suggestion
- `selectionMenuChild` - a boolean (defaults to `false`), whether the parent is a `selectionMenu` or not (used for events)
```

```html
<KDropdownItem :item="{ label: 'Leave the page', to: { path: '/' } }" />
```

There are 3 primary item types:

> - `router-link`

Reviewer: ```suggestion
- `link`
```

In the actual `KDropdownItem` component the type is `link`, not `router-link`

  - the generic type generated using the `items` prop on `KDropdownMenu`

>   - the generic type generated using the `item` prop on `KDropdownItem`

Reviewer: I think one of these lines was  duplicated on accident and is not needed

Author: Read closer you grasshopper. Two different props from two different components.

        class="d-inline-block"
      >
        <a
          href="http://www.google.com"
          rel="noopener"
          target="_blank"
        >
          External link
          <KIcon
            icon="externalLink"
            size="12"
            class="ml-2"
          />
        </a>
      </KDropdownItem>
    </template>
  </KDropdownMenu>
</div>

```html
<KDropdownMenu label="Variety">
  <template #items>
    <KDropdownItem :item="{ label: 'You are here', to: { path: '/components/dropdown-menu.html' } }" />
    <KDropdownItem @click="clickHandler">
      A button
    </KDropdownItem>
    <KDropdownItem
      disabled
      @click="clickHandler"
    >
      Disabled button
    </KDropdownItem>
    <KDropdownItem
      has-divider
      is-dangerous
      class="d-inline-block"
    >
      <a
        href="http://www.google.com"
        rel="noopener"
        target="_blank"
      >
        External link
        <KIcon
          icon="externalLink"
          size="12"
          class="ml-2"
        />
      </a>
    </KDropdownItem>
  </template>
</KDropdownMenu>
```

### Events

| Event     | Description             |
| :-------- | :------------------ |

> | `clicked` | Fires when a menu item is clicked |

Reviewer: Event name
```suggestion
| `click` | Fires when a menu item is clicked |
```

| `changed` | Fires when items with `selectionMenuChild` prop are clicked; returns `selectedItem` Object or null |

> 

Reviewer: We also need to document the `toggleDropdown` event

Author: Fixed

# Dropdown Menu

**KDropdownMenu** is a button (or any slotted content) that is clicked to trigger a menu popover beneath it.

<KDropdownMenu label="Documentation" :items="deepClone(defaultItemsUnselected)" />

## Props

### label

The label for the menu.

### items

An array of item objects containing a `label` property and other optional properties which will render a menu of [`KDropdownItems`](#KDropdownItem) .

```html
<KDropdownMenu
  label="Documentation"
  :items="[
    { label: 'Props', to: { path: '/components/dropdown-menu.html#props' } },
    { label: 'Slots', to: { path: '/components/dropdown-menu.html#slots' } },
    { label: 'Top', to: { path: '/components/dropdown-menu.html' } }
  ]"
/>
```

### appearance

Use this prop to specify the display style for the dropdown menu. Can be either `menu` (default) or `selectionMenu`.

> The `menu` style is the standard you have seen in the example above. Uses a standard `primary` `KButton` with hover state over items and no notion of "selection".

Author: Done

      </KDropdownItem>
    </template>
  </KDropdownMenu>
</div>

```html
<KDropdownMenu
  :label="selectedItem.label"
  appearance="selectionMenu"
  @change="(selection) => handleChange(selection)"
>
  <template #items="{ handleSelection }">
    <KDropdownItem
      v-for="item in menuItems"
      :selected="selectedItem.value === item.value"
      @click="handleSelection(item)"
    >
      {{ item.label }}
    </KDropdownItem>
  </template>
</KDropdownMenu>

<script>
export default {
  data() {
    return {
      selectedItem: {
        label: 'Select an item',
        value: ''
      },
      menuItems: [{
        label: 'US (United States)',
        value: 'us'
      },
      {
        label: 'FR (France)',
        value: 'fr'
      }]
    }
  },
  methods: {
    handleChange (item) {
      this.selectedItem = item
      this.$toaster.open(`${item.label} clicked!`)
    }
  }
}
</script>
```

### showCaret

Use this prop if you would like the trigger button to display the caret.

<KDropdownMenu label="Documentation" :items="deepClone(defaultItemsUnselected)" show-caret />

```html
<KDropdownMenu
  label="Documentation"
  :items="items"
  show-caret
/>
```

### width

> The width of the dropdown body. Currently we support numbers (will be converted to `px`), `auto`, and percentages for width.

Author: Fixed
